codec-http/src/test/java/io/netty5/handler/codec/http/headers/LegacyCookieParsingTest.java

```diff
@@ -0,0 +1,434 @@
+ * copy of the License at:
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.netty5.handler.codec.http.headers;
+
+import io.netty5.handler.codec.DateFormatter;
+import org.hamcrest.MatcherAssert;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.util.Calendar;
+import java.util.Date;
+import java.util.TimeZone;
+
+import static io.netty5.handler.codec.http.HttpHeaderNames.COOKIE;
+import static io.netty5.handler.codec.http.HttpHeaderNames.SET_COOKIE;
+import static io.netty5.handler.codec.http.headers.HttpSetCookie.SameSite.None;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.hamcrest.CoreMatchers.instanceOf;
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/**
+ * Tests from the old Netty 4.1 ClientCookieDecoder and ServerCookieDecoder.
+ * Those decoders were tested on cookie formats for older RFCs.
+ * We recreate those tests here, to verify that we're still compatible with those formats.
+ */
+public class LegacyCookieParsingTest {
+    HttpHeaders headers;
+
+    @BeforeEach
+    void createHeaders() {
+        headers = HttpHeaders.newHeaders(false);
+    }
+
+    // Server cookie (COOKIE header) decoder tests
+    @Test
+    void decodingSingleServerCookie() {
+        headers.add(COOKIE, "myCookie=myValue");
+        assertThat(headers.getCookies()).hasSize(1);
+        assertThat(headers.getCookies()).containsExactly(new DefaultHttpCookiePair("myCookie", "myValue"));
+    }
+
+    @Test
+    void decodingMultipleServerCookies() {
+        String c1 = "myCookie=myValue;";
+        String c2 = "myCookie2=myValue2;";
+        String c3 = "myCookie3=myValue3;";
+        headers.add(COOKIE, c1 + c2 + c3);
+        assertThat(headers.getCookies()).hasSize(3);
+        assertThat(headers.getCookies()).containsExactly(
+                new DefaultHttpCookiePair("myCookie", "myValue"),
+                new DefaultHttpCookiePair("myCookie2", "myValue2"),
+                new DefaultHttpCookiePair("myCookie3", "myValue3"));
+    }
+
+    @Test
+    void decodingMultipleSameNamedServerCookies() {
+        String c1 = "myCookie=myValue;";
+        String c2 = "myCookie=myValue2;";
+        String c3 = "myCookie=myValue3;";
```

Comment:
Interesting… this would probably fail validation when parsing SET_COOKIEs, but it doesn't for COOKIEs.

Reply:
It passes because of the `- 2` here: https://github.com/netty/netty/blob/main/codec-http/src/main/java/io/netty5/handler/codec/http/headers/HeaderUtils.java#L453
There's otherwise no check that we had a semi-colon with nothing after it. Since there are no more COOKIE headers, the `advanceCookieHeader()` will set the `current` field to `null`, and the iteration stops there.

Reply:
Looks like there are more issues hiding in this corner. Working on another commit.

Reply:
@idelpivnitskiy take a look at the latest commit https://github.com/netty/netty/pull/12833/commits/bcd88b10de1393e599d725d1acc803cc4f3da250